### PR TITLE
CPLP-1634 notifications

### PR DESCRIPTION
**Pull request by Phil91**
In this PR [CPLP-1634](https://jira.catena-x.net/browse/CPLP-1634) is resolved.

Please also review the [DB-PR](https://github.com/tractusx-team-portal-onboarding/devenv/pull/170)

### CODE REVIEW

File: src/portalbackend/PortalBackend.PortalEntities/Extensions/NotificationTopicExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            NotificationTypeId.WELCOME_APP_MARKETPLACE => NotificationTopicId.INFO,
+            NotificationTypeId.ACTION => NotificationTopicId.ACTION,
+            NotificationTypeId.APP_SUBSCRIPTION_REQUEST => NotificationTopicId.ACTION,
+            NotificationTypeId.SERVICE_REQUEST => NotificationTopicId.ACTION,
+            NotificationTypeId.APP_SUBSCRIPTION_ACTIVATION => NotificationTopicId.OFFER,
+            NotificationTypeId.APP_RELEASE_REQUEST => NotificationTopicId.OFFER,
+            NotificationTypeId.SERVICE_ACTIVATION => NotificationTopicId.OFFER,
+            _ => throw new ArgumentOutOfRangeException(nameof(typeId), typeId, "No NotificationTopicId defined for the given type")
+        };
+}
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.PortalEntities/Extensions/NotificationTopicExtensions.cs:46]: newline at end of file ;-)

File: src/portalbackend/PortalBackend.PortalEntities/Extensions/NotificationTopicExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+using Org.CatenaX.Ng.Portal.Backend.PortalBackend.PortalEntities.Enums;
+
+namespace Org.CatenaX.Ng.Portal.Backend.PortalBackend.PortalEntities.Extensions;
+
+public static class NotificationTopicExtensions
+{
+    public static NotificationTopicId GetNotificationTopic(this NotificationTypeId typeId) =>
+        typeId switch
+        {
+            NotificationTypeId.INFO => NotificationTopicId.INFO,
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.PortalEntities/Extensions/NotificationTopicExtensions.cs:30]: @jjeroch @Phil91 I suggest we should relocate this mapping into the database, entity name like 'NotificationTypeAssignedTopic'. Then the query would join this table to do the grouping.
Having this mapping in code implies that any later change would result in notifications that were created before the change would not longer comply with this mapping. When the mapping is done in the database exisiting notification records would remain unchanged while the mapped topic would change with the change of the mapping-table.

File: tests/portalbackend/PortalBackend.DBAccess.Tests/NotificationRepositoryTests.cs [hunk 3]
```diff
@@ -25,8 +25,10 @@
 using Org.CatenaX.Ng.Portal.Backend.PortalBackend.DBAccess.Tests.Setup;
 using Org.CatenaX.Ng.Portal.Backend.PortalBackend.PortalEntities.Entities;
 using Org.CatenaX.Ng.Portal.Backend.PortalBackend.PortalEntities.Enums;
-using Org.CatenaX.Ng.Portal.Backend.Tests.Shared.TestSeeds;
 using FluentAssertions;
+using Microsoft.EntityFrameworkCore;
```
Comment (ntruchsess) [inline tests/portalbackend/PortalBackend.DBAccess.Tests/NotificationRepositoryTests.cs:29]: keep usings in alphabetical order to improve readability.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/notification/Notification.Service/Models/NotificationSorting.cs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+namespace Notification.Service.Models;
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/notification/Notification.Service/Models/NotificationSorting.cs:21]: done

File: src/notification/Notification.Service/BusinessLogic/NotificationBusinessLogic.cs [hunk 5]
```diff
@@ -132,4 +173,4 @@ private async Task CheckNotificationExistsAndIamUserIsReceiver(Guid notification
             throw new ForbiddenException($"iamUserId {iamUserId} is not the receiver of the notification");
         }
     }
-}
+}
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/notification/Notification.Service/BusinessLogic/NotificationBusinessLogic.cs:176]: done

File: src/portalbackend/PortalBackend.PortalEntities/Extensions/NotificationTopicExtensions.cs (hunk 1 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.PortalEntities/Extensions/NotificationTopicExtensions.cs:46]: done

File: src/notification/Notification.Service/Models/NotificationSorting.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/notification/Notification.Service/Models/NotificationSorting.cs:21]: cannot see - any outstanding commits ready to push?

File: src/portalbackend/PortalBackend.DBAccess/Repositories/NotificationRepository.cs [hunk 6]
```diff
@@ -99,18 +102,26 @@ public Notification AttachAndModifyNotification(Guid notificationId, Action<Noti
...
+                    .Count(notification => !isRead.HasValue || notification.IsRead == isRead.Value)))
             .SingleOrDefaultAsync();
 
+    /// <inheritdoc />
+    public Task<Dictionary<(bool, NotificationTopicId), int>> GetCountDetailsForUserAsync(string iamUserId) => _dbContext.Notifications
+            .AsNoTracking()
+            .Where(not => not.Receiver!.IamUser!.UserEntityId == iamUserId)
+            .GroupBy(not => new {not.IsRead, not.NotificationTopicId})
+            .Select(not => new { Key = new ValueTuple<bool, NotificationTopicId>(not.Key.IsRead, not.Key.NotificationTopicId), Count = not.Count()})
+            .ToDictionaryAsync(x => x.Key, x => x.Count);
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/NotificationRepository.cs:117]: i like that. changed.

File: src/notification/Notification.Service/BusinessLogic/NotificationBusinessLogic.cs [hunk 7]
```diff
@@ -99,6 +127,19 @@ public async Task<int> GetNotificationCountAsync(string iamUserId, bool? isRead)
         return result.Count;
     }
 
+    /// <inheritdoc />
+    public async Task<NotificationCountDetails> GetNotificationCountDetailsAsync(string iamUserId)
+    {
+        var details = await _portalRepositories.GetInstance<INotificationRepository>().GetCountDetailsForUserAsync(iamUserId).ConfigureAwait(false);
+        var unreadNotifications = details.Where(x => !x.Key.IsRead);
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/notification/Notification.Service/BusinessLogic/NotificationBusinessLogic.cs:134]: changed

File: src/portalbackend/PortalBackend.PortalEntities/Extensions/NotificationTopicExtensions.cs (hunk 2 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.PortalEntities/Extensions/NotificationTopicExtensions.cs]: added the new mapping entity

File: src/portalbackend/PortalBackend.DBAccess/Repositories/NotificationRepository.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/NotificationRepository.cs]: need to check it tomorrow, currently getting an error from the query

[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/NotificationRepository.cs]: fixed

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**